### PR TITLE
feat: Add CredentialsProvider to Publisher and Subscriber settings

### LINKED ISSUE #472: Make it easier to use non defaults CredentialsProvider
**Is your feature request related to a problem? Please describe.**
To be able to set a CredentialsProvider on a SubscriberSettings, you need to set the credentials on a `Supplier<SubscriberServiceClient>`, a `Supplier<CursorServiceClient>`  and a `PartitionAssignmentServiceClient`

**Describe the solution you'd like**
Be able to set `CredentialsProvider` on the `SubscriberSettings` directly.

### PR DESCRIPTION
**Pull request by dpcollins-google**
These objects require more than one client and are annoying to construct correctly. Add a commonly needed setting to the top level here.

Fixes #472

### CODE REVIEW

File: google-cloud-pubsublite/clirr-ignored-differences.xml [hunk 1]
```diff
@@ -1,15 +1,11 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <!-- see http://www.mojohaus.org/clirr-maven-plugin/examples/ignored-differences.html -->
 <differences>
-  <!-- Removed not needed API -->
+  <!-- Added abstract method to AutoValue.Builder class (Always okay) -->
```
Comment (manuelmenzella-google) [inline google-cloud-pubsublite/clirr-ignored-differences.xml:4]: Can you instead add a specific exclusion for what you are adding? Not all changes to Builders will be adding abstract methods. Will this trigger on removing methods?

**Review by manuelmenzella-google — APPROVED** [review]
LGTM if my comment is addressed.

File: google-cloud-pubsublite/clirr-ignored-differences.xml (hunk 1 above)
Comment (dpcollins-google) [reply google-cloud-pubsublite/clirr-ignored-differences.xml:4]: This specifically triggers on adding methods to nested classes named "Builder". I think thats specific enough. (code 7013 is added abstract method to abstract class)

**Review by elharo — CHANGES_REQUESTED** [review]

File: google-cloud-pubsublite/clirr-ignored-differences.xml (hunk 1 above)
Comment (dpcollins-google) [reply google-cloud-pubsublite/clirr-ignored-differences.xml:4]: I disagree. Addition of a method to a Builder class, which is only overridden in an autogenerated AutoValue file, would not trigger a major version bump as it is not a breaking API surface change.

**Comment by suztomo** [comment]
@dpcollins-google I'm thinking to upgrade google-cloud-pubsublite dependency in Apache Beam (Google Dataflow) from 0.7.0 to 0.13.2. It has breaking changes including this PR475 ([compilation errors](https://issues.apache.org/jira/browse/BEAM-11010?focusedCommentId=17325144&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17325144)).  Before I try to fix the code, do you happen to know anyone working on Beam's pubsublite dependency upgrade already?

File: google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/cloudpubsub/SubscriberSettings.java [hunk 2]
```diff
@@ -132,67 +189,101 @@ public abstract Builder setSubscriberServiceClientSupplier(
...
+              .setSubscriptionPath(subscriptionPath())
+              .setServiceClient(newSubscriberServiceClient(partition));
+
+      Committer wireCommitter =
+          CommitterSettings.newBuilder()
+              .setSubscriptionPath(subscriptionPath())
+              .setPartition(partition)
+              .setServiceClient(newCursorServiceClient())
+              .build()
+              .instantiate();
```
Comment (suztomo) [inline google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/cloudpubsub/SubscriberSettings.java:242]: Memo for myself. This seems the way to replace CommitterBuilder:

```
        CommitterBuilder.Builder wireCommitterBuilder = CommitterBuilder.newBuilder();
        wireCommitterBuilder.setSubscriptionPath(canonicalPath);
        cursorServiceClientSupplier()
            .ifPresent(supplier -> wireCommitterBuilder.setServiceClient(supplier.get()));
        wireCommitterBuilder.setPartition(partition);
```

File: google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/cloudpubsub/PublisherSettings.java [hunk 3]
```diff
@@ -125,16 +188,12 @@ Publisher instantiate() throws ApiException {
                   SinglePartitionPublisherBuilder.Builder singlePartitionBuilder =
                       underlyingBuilder()
                           .setBatchingSettings(batchingSettings)
-                          .setContext(PubsubContext.of(FRAMEWORK))
```
Comment (suztomo) [inline google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/cloudpubsub/PublisherSettings.java:128]: Beam also touches this removed setContext method. Should it just remove the usage?

File: google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/internal/wire/PartitionCountWatchingPublisher.java [hunk 4]
```diff
@@ -94,11 +98,13 @@ public void stop() {
...
-        settings.configWatcherFactory().newWatcher(this::handleConfig);
+  PartitionCountWatchingPublisher(
+      PartitionPublisherFactory publisherFactory,
+      RoutingPolicy.Factory policyFactory,
+      PartitionCountWatcher.Factory configWatcherFactory) {
+    this.publisherFactory = publisherFactory;
+    this.policyFactory = policyFactory;
+    PartitionCountWatcher configWatcher = configWatcherFactory.newWatcher(this::handleConfig);
     addServices(configWatcher);
   }
```
Comment (suztomo) [inline google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/internal/wire/PartitionCountWatchingPublisher.java:109]: Memo for myself. This constructor is now invisible but used by Beam
https://github.com/apache/beam/blob/243128a8fc52798e1b58b0cf1a271d95ee7aa241/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsublite/Publishers.java#L46

File: pubsublite-beam-io/src/main/java/com/google/cloud/pubsublite/beam/PublisherOptions.java [hunk 5]
```diff
@@ -53,20 +62,38 @@ public boolean usesCache() {
     return clientSupplier() == null;
   }
 
+  private PublisherServiceClient newServiceClient(Partition partition) throws ApiException {
+    if (clientSupplier() != null) return clientSupplier().get();
+    PublisherServiceSettings.Builder settingsBuilder = PublisherServiceSettings.newBuilder();
+    settingsBuilder =
+        addDefaultMetadata(
+            PubsubContext.of(FRAMEWORK),
```
Comment (suztomo) [inline pubsublite-beam-io/src/main/java/com/google/cloud/pubsublite/beam/PublisherOptions.java:70]: This sets the context now.

File: google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/cloudpubsub/PublisherSettings.java (hunk 3 above)
Comment (suztomo) [reply google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/cloudpubsub/PublisherSettings.java:128]: This has been moved to newServiceClient method. (Thanks @dpcollins-google !)

**Pull request merged**